### ISSUE TITLE
Sometimes a cell with a loop runs forever

### ISSUE THREAD

**Issue by MichalKotecki** [googlecolab/colabtools#1700]
Hi, I need to resolve this for my thesis. My life depends on it.

I have a dataset of 10K colorful images of size 128x128 pixels. I uploaded the whole dataset to one folder in my Google Drive. My problem is most of the times iterating over dataset entries (for training and also showing some examples entries) does not work. Usually the cell with the iteration loop just runs forever and it does not event go to the first line of code inside the loop. A couple of times, after a few minutes, it did work and started the loop. If the loop starts there are no more problems with it. This is very frustrating, because I cannot train the network to finish my thesis. 

Could you please tell me what is the problem exactly and show some workarounds?

I load the dataset like so:

`my_dataset = tf.keras.preprocessing.image_dataset_from_directory(
    "/content/gdrive/My Drive/ColabDatasets/sm/" ,
    color_mode='rgb',
    image_size=(128,128),
    batch_size=256,
    label_mode=None
    )
`

Then I normalize each image:

`def normalize_single_image(image):
  return (tf.cast(image, tf.float32) - 127.5) / 127.5

my_dataset = my_dataset.map(normalize_single_image,
                                                  num_parallel_calls=tf.data.experimental.AUTOTUNE)
`

So far everything works and my_dataset is: 
`<ParallelMapDataset shapes: (None, 128, 128, 3), types: tf.float32>`

And here is the loop that causes all my headache:

`for batch in my_dataset.as_numpy_iterator():
	# Most of the times code inside here is not executed but the cell runs forever`

I use as_numpy_iterator(), because I need to display some images like so:

`index = 1
for batch in my_dataset.take(9).as_numpy_iterator():
  plt.subplot(3, 3, index)
  plt.axis('off')
  plt.imshow((batch[0] * 127.5 + 127.5).astype('uint8'))
  index += 1
`

**Comment by Suttsy3** [comment]
Where do you define "cartoon_faces_dataset"? Is that a curated training dataset from TensorFlow, or is it supposed to be a subset of your "my_dataset" array?

**Comment by MichalKotecki** [comment]
> Where do you define "cartoon_faces_dataset"? Is that a curated training dataset from TensorFlow, or is it supposed to be a subset of your "my_dataset" array?

It is actually the same thing as "my_dataset". I editted my question. I am sorry for the confussion.

**Comment by Suttsy3** [comment]
Have you added debug statements to evaluate if the loop is actually iterating? 

I ask because this seems like more of a loop definition or a dataset shape issue than a runtime issue in colab. The first value in your dataset shape array being None-type is likely throwing off the iteration on your For loop. Have you tried posting to TensorFlow's help boards or browsed Stack Overflow for similar issues?

Separately, Is there a reason you're forcing the training through a for loop instead of leveraging Keras's API model.fit?
